File: src/main/java/com/luminar/nova/lexer/Lexer.java
package com.luminar.nova.lexer;

import java.util.ArrayList;
import java.util.List;

public final class Lexer {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private final List<String> errors = new ArrayList<>();
    private int start = 0, current = 0, line = 1;

    public Lexer(String source) { this.source = source; }

    public List<Token> tokenize() {
        while (!isAtEnd()) { start = current; scanToken(); }
        tokens.add(new Token(TokenType.EOF, "", null, line));
        return tokens;
    }

    public boolean hasErrors()       { return !errors.isEmpty(); }
    public List<String> getErrors()  { return List.copyOf(errors); }

    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(' -> emit(TokenType.LPAREN);
            case ')' -> emit(TokenType.RPAREN);
            case '{' -> emit(TokenType.LBRACE);
            case '}' -> emit(TokenType.RBRACE);
            case '[' -> emit(TokenType.LBRACKET);
            case ']' -> emit(TokenType.RBRACKET);
            case ',' -> emit(TokenType.COMMA);
            case ';' -> emit(TokenType.SEMICOLON);
            case ':' -> emit(TokenType.COLON);
            case '?' -> emit(TokenType.QUESTION);
            case '.' -> { if (match('.')) emit(TokenType.DOT_DOT); else emit(TokenType.DOT); }
            case '+' -> { if (match('+')) emit(TokenType.PLUS_PLUS); else if (match('=')) emit(TokenType.PLUS_EQ);   else emit(TokenType.PLUS);  }
            case '-' -> { if (match('-')) emit(TokenType.MINUS_MINUS); else if (match('>')) emit(TokenType.ARROW); else if (match('=')) emit(TokenType.MINUS_EQ);  else emit(TokenType.MINUS); }
            case '*' -> { if (match('*')) emit(TokenType.STAR_STAR); else if (match('=')) emit(TokenType.STAR_EQ);   else emit(TokenType.STAR);  }
            case '/' -> {
                if      (match('/')) { while (!isAtEnd() && peek() != '\n') advance(); }
                else if (match('*')) blockComment();
                else if (match('=')) emit(TokenType.SLASH_EQ);
                else                 emit(TokenType.SLASH);
            }
            case '%' -> { if (match('=')) emit(TokenType.PERCENT_EQ); else emit(TokenType.PERCENT); }
            case '!' -> { if (match('=')) emit(TokenType.BANG_EQ);    else emit(TokenType.BANG);    }
            case '=' -> { if (match('=')) emit(TokenType.EQ_EQ);      else emit(TokenType.EQ);      }
            case '<' -> { if (match('=')) emit(TokenType.LESS_EQ);    else emit(TokenType.LESS);    }
            case '>' -> { if (match('=')) emit(TokenType.GREATER_EQ); else emit(TokenType.GREATER); }
            case '&' -> { if (match('&')) emit(TokenType.AMP_AMP);    else errors.add("Line " + line + ": Expected '&&'"); }
            case '|' -> { if (match('|')) emit(TokenType.PIPE_PIPE);   else errors.add("Line " + line + ": Expected '||'"); }
            case ' ', '\r', '\t' -> {}
            case '\n' -> line++;
            case '"'  -> scanString('"');
            case '\'' -> scanString('\'');
            default -> {
                if (Character.isDigit(c))     scanNumber();
                else if (isIdentStart(c))     scanIdent();
                else errors.add("Line " + line + ": Unexpected character '" + c + "'");
            }
        }
    }

    private void scanString(char quote) {
        var sb = new StringBuilder();
        while (!isAtEnd() && peek() != quote) {
            char ch = advance();
            if (ch == '\n') { line++; sb.append('\n'); continue; }
            if (ch == '\\') {
                char esc = advance();
                sb.append(switch (esc) {
                    case 'n'  -> '\n'; case 't' -> '\t'; case 'r' -> '\r';
                    case '\\' -> '\\'; case '"' -> '"'; case '\'' -> '\'';
                    default   -> esc;
                });
            } else sb.append(ch);
        }
        if (isAtEnd()) { errors.add("Line " + line + ": Unterminated string"); return; }
        advance();
        emitLit(TokenType.STRING, sb.toString());
    }

    private void scanNumber() {
        boolean isFloat = false;
        while (!isAtEnd() && (Character.isDigit(peek()) || peek() == '_')) advance();
        if (!isAtEnd() && peek() == '.' && current + 1 < source.length()
                && Character.isDigit(source.charAt(current + 1))) {
            isFloat = true; advance();
            while (!isAtEnd() && Character.isDigit(peek())) advance();
        }
        String raw = source.substring(start, current).replace("_", "");
        emitLit(TokenType.NUMBER, isFloat ? Double.parseDouble(raw) : Long.parseLong(raw));
    }

    private void scanIdent() {
        while (!isAtEnd() && isIdentPart(peek())) advance();
        String word = source.substring(start, current);
        TokenType type = TokenType.fromKeyword(word);
        Object lit = switch (type) { case TRUE -> Boolean.TRUE; case FALSE -> Boolean.FALSE; case NULL -> null; default -> null; };
        emitLit(type, lit);
    }

    private void blockComment() {
        int depth = 1;
        while (!isAtEnd() && depth > 0) {
            if (peek() == '\n')                            { line++; }
            if (peek() == '/' && peekNext() == '*')        { depth++; advance(); }
            else if (peek() == '*' && peekNext() == '/')   { depth--; advance(); }
            advance();
        }
    }

    private void emit(TokenType t)             { tokens.add(new Token(t, source.substring(start, current), null, line)); }
    private void emitLit(TokenType t, Object v){ tokens.add(new Token(t, source.substring(start, current), v, line)); }
    private char advance()                     { return source.charAt(current++); }
    private boolean match(char e)              { if (isAtEnd() || source.charAt(current) != e) return false; current++; return true; }
    private char peek()                        { return isAtEnd() ? '\0' : source.charAt(current); }
    private char peekNext()                    { return current + 1 >= source.length() ? '\0' : source.charAt(current + 1); }
    private boolean isAtEnd()                  { return current >= source.length(); }
    private boolean isIdentStart(char c)       { return Character.isLetter(c) || c == '_'; }
    private boolean isIdentPart(char c)        { return Character.isLetterOrDigit(c) || c == '_'; }
}

File: nova/lexer/lexer.py
from typing import List, Optional
from .token_types import TokenType, from_keyword
from .token import Token


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.tokens: List[Token] = []
        self.errors: List[str] = []
        self.start = 0
        self.current = 0
        self.line = 1

    def tokenize(self) -> List[Token]:
        while not self._is_at_end():
            self.start = self.current
            self._scan_token()
        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens

    def has_errors(self) -> bool:
        return len(self.errors) > 0

    def get_errors(self) -> List[str]:
        return self.errors.copy()

    def _scan_token(self):
        c = self._advance()
        if c == '(':
            self._emit(TokenType.LPAREN)
        elif c == ')':
            self._emit(TokenType.RPAREN)
        elif c == '{':
            self._emit(TokenType.LBRACE)
        elif c == '}':
            self._emit(TokenType.RBRACE)
        elif c == '[':
            self._emit(TokenType.LBRACKET)
        elif c == ']':
            self._emit(TokenType.RBRACKET)
        elif c == ',':
            self._emit(TokenType.COMMA)
        elif c == ';':
            self._emit(TokenType.SEMICOLON)
        elif c == ':':
            self._emit(TokenType.COLON)
        elif c == '?':
            self._emit(TokenType.QUESTION)
        elif c == '.':
            if self._match('.'):
                self._emit(TokenType.DOT_DOT)
            else:
                self._emit(TokenType.DOT)
        elif c == '+':
            if self._match('+'):
                self._emit(TokenType.PLUS_PLUS)
            elif self._match('='):
                self._emit(TokenType.PLUS_EQ)
            else:
                self._emit(TokenType.PLUS)
        elif c == '-':
            if self._match('-'):
                self._emit(TokenType.MINUS_MINUS)
            elif self._match('>'):
                self._emit(TokenType.ARROW)
            elif self._match('='):
                self._emit(TokenType.MINUS_EQ)
            else:
                self._emit(TokenType.MINUS)
        elif c == '*':
            if self._match('*'):
                self._emit(TokenType.STAR_STAR)
            elif self._match('='):
                self._emit(TokenType.STAR_EQ)
            else:
                self._emit(TokenType.STAR)
        elif c == '/':
            if self._match('/'):
                while not self._is_at_end() and self._peek() != '\n':
                    self._advance()
            elif self._match('*'):
                self._block_comment()
            elif self._match('='):
                self._emit(TokenType.SLASH_EQ)
            else:
                self._emit(TokenType.SLASH)
        elif c == '%':
            if self._match('='):
                self._emit(TokenType.PERCENT_EQ)
            else:
                self._emit(TokenType.PERCENT)
        elif c == '!':
            if self._match('='):
                self._emit(TokenType.BANG_EQ)
            else:
                self._emit(TokenType.BANG)
        elif c == '=':
            if self._match('='):
                self._emit(TokenType.EQ_EQ)
            else:
                self._emit(TokenType.EQ)
        elif c == '<':
            if self._match('='):
                self._emit(TokenType.LESS_EQ)
            else:
                self._emit(TokenType.LESS)
        elif c == '>':
            if self._match('='):
                self._emit(TokenType.GREATER_EQ)
            else:
                self._emit(TokenType.GREATER)
        elif c == '&':
            if self._match('&'):
                self._emit(TokenType.AMP_AMP)
            else:
                self.errors.append(f"Line {self.line}: Expected '&&'")
        elif c == '|':
            if self._match('|'):
                self._emit(TokenType.PIPE_PIPE)
            else:
                self.errors.append(f"Line {self.line}: Expected '||'")
        elif c in (' ', '\r', '\t'):
            pass
        elif c == '\n':
            self.line += 1
        elif c == '"' or c == "'":
            self._scan_string(c)
        else:
            if c.isdigit():
                self._scan_number()
            elif self._is_ident_start(c):
                self._scan_ident()
            else:
                self.errors.append(f"Line {self.line}: Unexpected character '{c}'")

    def _scan_string(self, quote: str):
        sb = []
        while not self._is_at_end() and self._peek() != quote:
            ch = self._advance()
            if ch == '\n':
                self.line += 1
                sb.append('\n')
                continue
            if ch == '\\':
                esc = self._advance()
                if esc == 'n':
                    sb.append('\n')
                elif esc == 't':
                    sb.append('\t')
                elif esc == 'r':
                    sb.append('\r')
                elif esc == '\\':
                    sb.append('\\')
                elif esc == '"':
                    sb.append('"')
                elif esc == "'":
                    sb.append("'")
                else:
                    sb.append(esc)
            else:
                sb.append(ch)
        if self._is_at_end():
            self.errors.append(f"Line {self.line}: Unterminated string")
            return
        self._advance()
        self._emit_lit(TokenType.STRING, ''.join(sb))

    def _scan_number(self):
        is_float = False
        while not self._is_at_end() and (self._peek().isdigit() or self._peek() == '_'):
            self._advance()
        if (not self._is_at_end() and self._peek() == '.'
                and self.current + 1 < len(self.source)
                and self.source[self.current + 1].isdigit()):
            is_float = True
            self._advance()
            while not self._is_at_end() and self._peek().isdigit():
                self._advance()
        raw = self.source[self.start:self.current].replace('_', '')
        if is_float:
            value = float(raw)
        else:
            value = int(raw)
        self._emit_lit(TokenType.NUMBER, value)

    def _scan_ident(self):
        while not self._is_at_end() and self._is_ident_part(self._peek()):
            self._advance()
        word = self.source[self.start:self.current]
        token_type = from_keyword(word)
        literal = None
        if token_type == TokenType.TRUE:
            literal = True
        elif token_type == TokenType.FALSE:
            literal = False
        elif token_type == TokenType.NULL:
            literal = None
        self._emit_lit(token_type, literal)

    def _block_comment(self):
        depth = 1
        while not self._is_at_end() and depth > 0:
            if self._peek() == '\n':
                self.line += 1
            if self._peek() == '/' and self._peek_next() == '*':
                depth += 1
                self._advance()
            elif self._peek() == '*' and self._peek_next() == '/':
                depth -= 1
                self._advance()
            self._advance()

    def _emit(self, token_type: TokenType):
        lexeme = self.source[self.start:self.current]
        self.tokens.append(Token(token_type, lexeme, None, self.line))

    def _emit_lit(self, token_type: TokenType, value: object):
        lexeme = self.source[self.start:self.current]
        self.tokens.append(Token(token_type, lexeme, value, self.line))

    def _advance(self) -> str:
        ch = self.source[self.current]
        self.current += 1
        return ch

    def _match(self, expected: str) -> bool:
        if self._is_at_end() or self.source[self.current] != expected:
            return False
        self.current += 1
        return True

    def _peek(self) -> str:
        if self._is_at_end():
            return '\0'
        return self.source[self.current]

    def _peek_next(self) -> str:
        if self.current + 1 >= len(self.source):
            return '\0'
        return self.source[self.current + 1]

    def _is_at_end(self) -> bool:
        return self.current >= len(self.source)

    def _is_ident_start(self, c: str) -> bool:
        return c.isalpha() or c == '_'

    def _is_ident_part(self, c: str) -> bool:
        return c.isalnum() or c == '_'
